### PR TITLE
[SV] Move emitAsComment to SVAttributeAttr; remove SVAttributesAttr

### PR DESCRIPTION
**Pull request by fabianschuiki**
Move the `emitAsComments` flag from `SVAttributesAttr` (plural) to `SVAttributeAttr` (singular). Then remove `SVAttributesAttr` (plural) since it only contains a single array field after `emitAsComments` is gone.

Add add, remove, and read-modify-write functions to deal with SV attributes on operations more conveniently. Also enforce deduplicated SV attributes where easily possible. The new functions now allow for easy addition and removal of individual SV attributes.

Adapt ExportVerilog to handle the fact that emission as comments is now a per-attribute option. The emission code now can emit multiple `/*...*/` and `(*...*)` groups if needed, but tries to pack as many consecutive attributes with the same `emitAsComment` flag into one group.

### CODE REVIEW

**Review by uenoku — APPROVED** [review]
Can we add a test where an array contains multiple SV attributes including `emitAsComment` being true?

File: include/circt/Dialect/SV/SVAttributes.td [hunk 1]
```diff
@@ -53,42 +53,41 @@ def SVAttributeAttr : AttrDef<SVDialect, "SVAttribute"> {
...
+                 StringAttr::get(context, expression),
+                 BoolAttr::get(context, emitAsComment));
+    }]>,
+    AttrBuilder<(ins "::mlir::StringAttr":$name,
+                     "::mlir::StringAttr":$expression,
+                     CArg<"bool", "false">: $emitAsComment), [{
+      return get(context, name, expression,
+                 BoolAttr::get(context, emitAsComment));
+    }]>
+  ];
```
Comment (uenoku) [inline include/circt/Dialect/SV/SVAttributes.td:87]: It's unfortunate that MLIR cannot generate these ;)

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 2]
```diff
@@ -751,18 +751,38 @@ static IfOp findNestedElseIf(Block *elseBlock) {
...
+      os << "/* ";
+    else if (newContainer == InAttr)
+      os << "(* ";
+    currentContainer = newContainer;
+    return true;
+  };
+
+  for (auto attr : attrs.getAsRange<SVAttributeAttr>()) {
+    if (!openContainer(attr.getEmitAsComment().getValue() ? InComment : InAttr))
+      os << ", ";
```
Comment (uenoku) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:780]: nit: Can we insert a space when we open a new container unless it's not first element?

**Comment by dtzSiFive** [comment]
PP-enabled version, with some tests: https://github.com/dtzSiFive/circt/commits/fschuiki/sv-attributes-array .
Quick look at spec suggests the attribute "containers" are allowed to be split (attributes may be multi-line) -- do we know what tool compatibility story is here?

Idea is to put containers on own line if needed to fit, otherwise put each on their own line.
Within a container, break between attributes (don't break up a single `key = value` in case that matters) aligning to column the first starts at (see `pretty.mlir` for example output).

No breaking if emitting as part of an expression, but can tackle that if needed.  We currently intentionally ensure the space before emitting SV attributes on an expression has a non-breaking space before that, so kept that sort of behavior.

Anyway, if that sounds like how we want it take a look at that sort of approach :+1:, happy to help tweak.

File: include/circt/Dialect/SV/SVAttributes.td (hunk 1 above)
Comment (fabianschuiki) [reply include/circt/Dialect/SV/SVAttributes.td:87]: Yeah. Just listing the entire combinatorial explosion of param combinations isn't really a good approach. In Rust land there's the "builder pattern" that people use, where you create a helper struct for collecting fields in multiple steps:
```c++
struct Builder {
  Builder &name(StringRef arg) { return name(arg); }
  Builder &name(StringAttr arg) { nameArg = StringAttr::get(arg); return *this; }
  Builder &expression(StringRef arg) { return expression(arg); }
  Builder &expression(StringAttr arg) { expressionArg = StringAttr::get(arg); return *this; }
  Builder &emitAsComment(bool arg = true) { return emitAsComment(arg); }
  Builder &emitAsComment(BoolAttr arg) { emitAsCommentArg = BoolAttr::get(arg); return *this; }
  SVAttributeAttr build() {
    return SVAttributeAttr::get(nameArg, expressionArg, emitAsCommentArg);
  }

private:
  StringAttr nameArg;
  StringAttr expressionArg;
  BoolAttr emitAsCommentArg;
};

// Usage:
Builder().name("bla").build()
Builder().name(strAttr).expression("foobar").emitAsComment(boolAttr).build()
```

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (fabianschuiki) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp:780]: Yeah great idea!

**Comment by dtzSiFive** [comment]
> Thanks a lot @dtzSiFive for the PP version! I'll just pull that into this PR if that's okay with you smiley

Yes please, and ofc/you're welcome.  Wasn't sure if this is the direction we wanted to go, so didn't want to just push on top.  But by all means :+1:.

**Comment by fabianschuiki** [comment]
That pretty printer is a thing of beauty. Donald Knuth approves.

**Pull request merged**